src/per/spi.cpp

```diff
@@ -8,126 +11,547 @@
+            baud_prescaler = SPI_BAUDRATEPRESCALER_32;
+            break;
+        case Config::BaudPrescaler::BAUDRATEPRESCALER_64:
+            baud_prescaler = SPI_BAUDRATEPRESCALER_64;
+            break;
+        case Config::BaudPrescaler::BAUDRATEPRESCALER_128:
+            baud_prescaler = SPI_BAUDRATEPRESCALER_128;
+            break;
+        case Config::BaudPrescaler::BAUDRATEPRESCALER_256:
+            baud_prescaler = SPI_BAUDRATEPRESCALER_256;
+            break;
+        default: return Result::ERR;
+    }
+
+    hspi_.Instance               = periph;
+    hspi_.Init.Mode              = mode;
+    hspi_.Init.Direction         = direction;
+    hspi_.Init.DataSize          = datasize;
+    hspi_.Init.CLKPolarity       = clock_polarity;
+    hspi_.Init.CLKPhase          = clock_phase;
+    hspi_.Init.NSS               = nss;
+    hspi_.Init.BaudRatePrescaler = baud_prescaler;
+    hspi_.Init.FirstBit          = SPI_FIRSTBIT_MSB;
+    hspi_.Init.TIMode            = SPI_TIMODE_DISABLE;
+    hspi_.Init.CRCCalculation    = SPI_CRCCALCULATION_DISABLE;
+    hspi_.Init.CRCPolynomial     = 0x0;
+    hspi_.Init.NSSPMode          = SPI_NSS_PULSE_ENABLE;
+    hspi_.Init.NSSPolarity       = SPI_NSS_POLARITY_LOW;
+    hspi_.Init.FifoThreshold     = SPI_FIFO_THRESHOLD_01DATA;
+    hspi_.Init.TxCRCInitializationPattern
         = SPI_CRC_INITIALIZATION_ALL_ZERO_PATTERN;
-    hspi1.Init.RxCRCInitializationPattern
+    hspi_.Init.RxCRCInitializationPattern
         = SPI_CRC_INITIALIZATION_ALL_ZERO_PATTERN;
-    hspi1.Init.MasterSSIdleness        = SPI_MASTER_SS_IDLENESS_00CYCLE;
-    hspi1.Init.MasterInterDataIdleness = SPI_MASTER_INTERDATA_IDLENESS_00CYCLE;
-    hspi1.Init.MasterReceiverAutoSusp  = SPI_MASTER_RX_AUTOSUSP_DISABLE;
-    hspi1.Init.MasterKeepIOState       = SPI_MASTER_KEEP_IO_STATE_DISABLE;
-    hspi1.Init.IOSwap                  = SPI_IO_SWAP_DISABLE;
-    if(HAL_SPI_Init(&hspi1) != HAL_OK)
+    hspi_.Init.MasterSSIdleness        = SPI_MASTER_SS_IDLENESS_00CYCLE;
+    hspi_.Init.MasterInterDataIdleness = SPI_MASTER_INTERDATA_IDLENESS_00CYCLE;
+    hspi_.Init.MasterReceiverAutoSusp  = SPI_MASTER_RX_AUTOSUSP_DISABLE;
+    hspi_.Init.MasterKeepIOState       = SPI_MASTER_KEEP_IO_STATE_DISABLE;
+    hspi_.Init.IOSwap                  = SPI_IO_SWAP_DISABLE;
+    if(HAL_SPI_Init(&hspi_) != HAL_OK)
     {
         Error_Handler();
+        return SpiHandle::Result::ERR;
     }
+
+    return SpiHandle::Result::OK;
 }
 
-void SpiHandle::BlockingTransmit(uint8_t* buff, size_t size)
+SpiHandle::Result SpiHandle::Impl::BlockingTransmit(uint8_t* buff, size_t size)
 {
-    HAL_SPI_Transmit(&hspi1, buff, size, 100);
+    if(HAL_SPI_Transmit(&hspi_, buff, size, 100) != HAL_OK)
+    {
+        return SpiHandle::Result::ERR;
+    }
+    return SpiHandle::Result::OK;
 }
 
-void HAL_SPI_MspInit(SPI_HandleTypeDef* spiHandle)
+SpiHandle::Result
+SpiHandle::Impl::PollReceive(uint8_t* buffer, uint16_t size, uint32_t timeout)
```

Comment:
Seems like the corresponding function in the header file is mising

Reply:
yep, fixed